### PR TITLE
lib: use explicit name imports in `lib/systems`

### PR DESCRIPTION
**Pull request by philiptaron**
## Description of changes

Using `with` at a top-level scope is [an anti-pattern](https://nix.dev/guides/best-practices#with-scopes). The tracking issue is https://github.com/NixOS/nixpkgs/issues/208242. This is a pure refactor: there is no functional change contained in this PR.

The tests found issues when I intentionally introduced them and obviously continue to pass.

Related: 
- https://github.com/NixOS/nixpkgs/pull/293901
- https://github.com/NixOS/nixpkgs/pull/294906

## Things done

- [x] Tested, as applicable:
  - [x] `nix-instantiate --strict --eval lib/tests/misc.nix`
  - [x] `lib/tests/modules.sh`
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: lib/systems/inspect.nix [hunk 1]
```diff
@@ -1,14 +1,64 @@
...
+
+  # These patterns depend on each other, so they're at the top level.
+  isBSD          = { kernel = { families = { inherit (kernelFamilies) bsd; }; }; };
+  isDarwin       = { kernel = { families = { inherit (kernelFamilies) darwin; }; }; };
+  isUnix         = [ isBSD isDarwin isLinux isSunOS isCygwin isRedox ];
+
+  isLinux        = { kernel = kernels.linux; };
+  isSunOS        = { kernel = kernels.solaris; };
+  isCygwin       = { kernel = kernels.windows; abi = cygnus; };
+  isRedox        = { kernel = kernels.redox; };
```
Comment (fricklerhandwerk) [inline lib/systems/inspect.nix:58]: Pulling these things apart makes it harder to have things in overview. I don't think `rec` per se is bad if used judiciously, and can in fact greatly improve clarity. That ["best practice" recommendation](https://nix.dev/guides/best-practices#recursive-attribute-set-rec) is an oversimplification and needs more work.

Comment (philiptaron) [reply lib/systems/inspect.nix:58]: I hear that. I find the exact semantics of `rec` really hard to articulate, but you're right that it's a separate-ish concern from `with`. Perhaps this is me suffering from the overeager schoolboy problem of trying to get the "extra" in.

Comment (philiptaron) [reply lib/systems/inspect.nix]: I've removed the commit that did this from this PR.

File: lib/systems/parse.nix [hunk 2]
```diff
@@ -166,7 +197,41 @@ rec {
   # Note: Since 22.11 the archs of a mode switching CPU are no longer considered
   # pairwise compatible. Mode switching implies that binaries built for A
   # and B respectively can't be executed at the same time.
-  isCompatible = a: b: with cpuTypes; lib.any lib.id [
+  isCompatible = let
```
Comment (fricklerhandwerk) [inline lib/systems/parse.nix:200]: This was a sensible use of `with`, and duplicating every occurrence of CPU types in this block is exclusively detrimental to maintainability IMO, especially now that you made all other symbols explicit.

Comment (philiptaron) [reply lib/systems/parse.nix:200]: I'll revert. Sorry for the failure to apply feedback here.

**Review by fricklerhandwerk — APPROVED** [review]

**Comment by Ericson2314** [comment]
Just a quibble (PR is fine), but most of it doesn't seems to be getting rid of `with`? I think explicit `lib.<name>` is fine? Sometimes "qualified name" style like that is easier to read because you don't need to remember what comes from where: "oh, it is calling some `lib` function" is immediately obvious.

**Comment by philiptaron** [comment]
> Just a quibble (PR is fine), but most of it doesn't seems to be getting rid of `with`? I think explicit `lib.<name>` is fine? Sometimes "qualified name" style like that is easier to read because you don't need to remember what comes from where: "oh, it is calling some `lib` function" is immediately obvious.

Following advice, I also preferred to inherit names from `lib` instead of `builtins` where there was a choice. I also preferred to inherit from `lib` in one place, rather than to have a mixture of `lib.thing` and `inherit (lib) thing`.

The PR did used to include other things and have a description referring to those other things, but I dropped them following feedback.

Still learning how to scope and describe refactor PRs that will get speedy acceptance and merging. This is one of my first ones from a few weeks ago. I appreciate your feedback!

**Comment by Ericson2314** [comment]
others might disagree with me on this, but I think I prefer `lib.<name>` unless it is used so many times `let inherit (lib) name; in ...` really makes things briefer.

**Comment by philiptaron** [comment]
> others might disagree with me on this, but I think I prefer `lib.<name>` unless it is used so many times `let inherit (lib) name; in ...` really makes things briefer.

I wonder if I can construct a re-writer that would follow this preference...

My preference is the opposite (as you can see) but that's mostly because I like consistency of the thorough variety ([the notable hob-goblin of small minds](https://www.goodreads.com/quotes/353571-a-foolish-consistency-is-the-hobgoblin-of-little-minds-adored)).

**Pull request merged**